sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpMessageExtensions.cs

```diff
@@ -37,39 +45,39 @@ private static byte[] GetByteArray(this Data data)
             }
         }
 
-        private static IEnumerable<byte[]> GetDataViaDataBody(AmqpMessage message)
+        public static IEnumerable<BinaryData> GetDataViaDataBody(this AmqpMessage message)
         {
             foreach (Data data in (message.DataBody ?? Enumerable.Empty<Data>()))
             {
                 byte[] bytes = data.GetByteArray();
                 if (bytes != null)
                 {
-                    yield return bytes;
+                    yield return BinaryData.FromBytes(bytes);
                 }
             }
         }
 
         // Returns via the out parameter the flattened collection of bytes.
         // A majority of the time, data will only contain 1 element.
         // The method is optimized for this situation to return the pre-existing array.
-        private static byte[] ConvertAndFlattenData(IEnumerable<byte[]> data)
+        public static BinaryData ConvertAndFlattenData(this IEnumerable<BinaryData> dataList)
         {
-            byte[] flattened = null;
+            ReadOnlyMemory<byte> flattened = default;
             List<byte> flattenedList = null;
             var dataCount = 0;
-            foreach (byte[] byteArray in data)
+            foreach (BinaryData data in dataList)
             {
                 // Only the first array is needed if it is the only valid array.
                 // This should be the case 99% of the time.
                 if (dataCount == 0)
                 {
-                    flattened = byteArray;
+                    flattened = data;
                 }
                 else
                 {
                     // We defer creating this list since this case will rarely happen.
-                    flattenedList ??= new List<byte>(flattened!);
-                    flattenedList.AddRange(byteArray);
+                    flattenedList ??= new List<byte>(flattened.ToArray()!);
+                    flattenedList.AddRange(data.ToBytes().ToArray());
```

Comment:
That's a lot of construction just to add something to the list (in both cases). That's why I kept everything as bytes to do this processing since it involved the least amount of construction. Then, after we do the processing, I made the proper type as necessary.

Reply:
Yes, I couldn't just use bytes since the AmqpAnnotatedMessage body is defined as BinaryData.

Reply:
Updated to use ArrayBufferWriter.